### ISSUE TITLE
[RFC]: Dual Batch Overlap (DBO) for Model Runner V2

### ISSUE THREAD

**Issue by specture724** [vllm-project/vllm#50738]
### Motivation.

Dual Batch Overlap hides all-to-all communication latency in DP+EP deployments by splitting a step's batch into two microbatches and interleaving their compute and comm phases across two CUDA streams. Today it exists only in the V1 model runner (`gpu_ubatch_wrapper.py` / `ubatching.py`), wired in as a `UBatchWrapper` that wraps the whole model and rebuilds attention metadata per microbatch after the fact.

Model Runner V2 (`vllm/v1/worker/gpu/`) is meant to eventually replace V1, but it has no DBO path. `_get_v2_model_runner_unsupported_features()` currently raises if `--enable-dbo` is combined with V2, so **DP+EP deployments that depend on DBO for throughput cannot move to V2**. That's a real migration blocker for exactly the multi-node MoE serving configurations V2 is aimed at.

Porting V1's implementation directly doesn't fit. V1's DBO is architecturally opportunistic: `split_attn_metadata` slices *already-built* full-batch attention metadata after the fact, allocating fresh tensors per microbatch per step. That's fine for V1's dynamic-shape eager/graph-per-shape model, but it's hostile to V2's core design — V2 dispatches per `BatchExecutionDescriptor` against a graph registry and reuses **persistent** input buffers across replays so tensor addresses stay stable. Naively porting V1's post-hoc splitting would mean allocating inside the hot path on every microbatched step, undermining the reason V2 exists.

This RFC proposes a V2-native DBO implementation that reuses the runner-agnostic model/kernel-side machinery (the `dbo_yield*` hooks in `fused_moe/modular_kernel.py`, DeepEP prepare/finalize, `mla_attention.py`, EPLB) unchanged, but replaces the runner glue with something that respects V2's persistent-buffer, explicit-dispatch model from the start.

**Status:** an eager-mode prototype (P0–P2) is implemented and measured on DP2/TP1/EP2 DeepSeek-V2-Lite. Overlap works — 69.5% of the all-to-all is now hidden behind compute, up from 11.4% on the baseline — and it converts to **+4.6% end-to-end throughput on large-prefill traffic**. It also *regresses decode by 2.25×*, for reasons that turn out to be structural rather than fixable. Numbers and analysis are in Measured results session; the gating question they raise is the main thing I'd like feedback on.

### Proposed Change.

DBO has always had two layers:

- **Model/kernel-side** (`ubatching.py`'s `UBatchContext`, the `dbo_yield*` hooks called from MoE/MLA/DeepEP code, EPLB overlap hooks): this is runner-agnostic — it just needs two CUDA streams, two forward contexts, and a CPU handoff protocol. **Reused unchanged for V2.**
- **Runner glue** (how the batch gets split, how attention metadata gets built per microbatch, how DP ranks agree to microbatch, how the split interacts with CUDA graphs): this is where V1 and V2 fundamentally differ and where new code is needed.

### Key design decisions

**1. `num_ubatches` as a first-class dispatch key.** `BatchExecutionDescriptor` (the struct V2 uses to key its CUDA-graph registry) gains a `num_ubatches: int = 1` field, and `_is_compatible` requires an exact match — a graph captured for N microbatches can only serve a batch split N ways. A microbatched step is just a descriptor with `num_ubatches=2` and `cg_mode=NONE`; it composes with the existing dispatch machinery instead of branching around it.

**2. Persistent per-microbatch buffers, sliced *before* metadata is built.** This is the main structural departure from V1. `InputBuffers` gets `num_ubatches`-many persistent `ubatch_query_start_loc` / `ubatch_seq_lens` buffers allocated once at init. Slicing a microbatch's `InputBatch` (`gpu/ubatch_utils.py::slice_input_batch`) writes into these fixed buffers via `torch.sub(...).clamp_()` in place, then attention metadata is built *from* the sliced batch — not sliced *out of* pre-built metadata. Everything except `query_start_loc` and `seq_lens` is a view of buffers the full batch already uses, so slicing costs no allocation, and buffer addresses are stable across steps, which is what CUDA graph replay requires (graph capture for microbatched shapes is P3, but the buffer layout is DBO-safe from P0).

One simplification over V1 falls out of this: V1 special-cases request boundaries and cudagraph padding as two separate branches in `split_attn_metadata`. Because the V2 buffers are sized to `num_tokens_after_padding` up front, both cases collapse into a single `.clamp_(0, num_tokens_padded)` — the clamp absorbs the padding for free.

**3. One attention metadata builder per microbatch.** V2 hardcoded `num_metadata_builders=1` in `init_attn_backend`. That is not compatible with DBO: microbatches build attention metadata concurrently on two threads, and some builders keep prepared metadata *on themselves* — MLA stores it on the prefill backend object, as `mla_attention.py` states outright. With a shared builder the two threads race on `_prefill_backend._prefill_metadata` and the engine dies on `assert self._prefill_metadata.chunked_context is not None`. So V2 allocates `num_ubatches` builders when ubatching is on, and `build_attn_metadata` / `prepare_attn` take a `ubatch_idx` to select among them. The `AttentionGroup` API already supported `ubatch_id`; V2 simply never wired it, and V1 has always sized builders by `num_ubatches`. Non-DBO paths are untouched (builder count stays 1, `ubatch_idx` defaults to 0).

**4. DP coordination extends the existing vote, fails safe.** DP ranks already all-reduce on CPU each step to agree on token counts and cudagraph mode. This gets one more row in the same tensor: each rank votes whether it *wants* to microbatch (`wants_ubatch`, based on `dbo_decode_token_threshold` / `dbo_prefill_token_threshold`). Microbatching only happens if **every** rank votes yes; any disagreement — including an idle rank that has nothing scheduled, or a rank whose token count is too small to fill both microbatches (`is_last_ubatch_empty`) — falls back to `num_ubatches=1` for that step. When it does engage, every rank is padded up to the largest rank's token count so all ranks can assume the others' microbatches are the same size. This trades a missed overlap opportunity for a guarantee that every rank takes the same path through the step.

**5. `UBatchRunner`: new V2-native orchestration, same threading primitive.** One thread per microbatch, `threading.Event` pairs for CPU handoff, a `threading.Barrier` for startup sync — this part follows V1's approach because it works and there's no reason to reinvent it. What's new is everything around it: per-microbatch `DPMetadata`/forward-context construction from the pre-sliced buffers, SM control (`create_sm_control_context`, promoted out of V1's wrapper into the already-shared `ubatch_utils.py` so both runners use one copy), and merging the per-microbatch outputs back into a single result (`IntermediateTensors` for PP-style outputs, tuples for EAGLE3 aux hidden states, or a plain tensor). Sampling still runs once over the merged batch — the sub-batches describe the forward pass only.

**6. Config gating stays conservative.** `VLLM_USE_V2_MODEL_RUNNER=1` must be set explicitly to get DBO on V2; if the env var is unset, DBO auto-selects V1 as before (V2 doesn't yet capture CUDA graphs for microbatched shapes — see P3 — so V1 keeps parity for now). Combining V2 DBO with LoRA, spec decode, PP, CP, multimodal/encoder-decoder, or hybrid models is rejected with a per-feature error rather than silently falling back, matching the existing `_get_v2_model_runner_unsupported_features` pattern.

### Stages

| Stage | Scope | Status |
|---|---|---|
| **P0** | Scaffolding: `num_ubatches` on the descriptor, persistent per-microbatch buffers on `InputBuffers`, microbatch slicing of `InputBatch`, no execution wiring yet. | **Done** |
| **P1** | Eager-mode DBO end-to-end: split-point policy, per-microbatch attention metadata + forward contexts, `UBatchRunner`, SM control, DP vote. Runs without CUDA graphs (`cg_mode=NONE`) for microbatched steps. | **Done** |
| **P2** | Per-ubatch metadata builders (decision 3); DP coordination edge cases: dummy/idle-rank runs join the same DP handshake as active ranks, skewed load across ranks, microbatches that can't be filled. | **Done** |
| **P3** | CUDA graph capture for microbatched shapes: needs a pre-capture hook so ubatch threads initialize cuBLAS before `torch.cuda.graph` opens (a known CUDA-graph/cuBLAS-lazy-init gotcha), and writing hidden states into the persistent buffer instead of `torch.cat`. This is what lets `cg_mode` be something other than `NONE` for microbatched steps. | Not started |
| **P4** | Perf validation and tuning. Overlap and throughput measured on DP2/EP2 (below). Still open: GSM8K accuracy parity vs V1, re-tuning the thresholds beyond the recommended config, a request-aligned split heuristic instead of a pure token-count split, and the workload-gating policy question. Flip the default gate once P3 lands and parity is shown. | **In progress** |

## Measured results

DeepSeek-V2-Lite (27 layers, 26 MoE, 64 routed experts, top-6), 2× H200 intranode, `DP2/TP1/EP2`, `deepep_high_throughput`, `VLLM_USE_V2_MODEL_RUNNER=1 --enforce-eager`. `vllm bench serve --request-rate inf --ignore-eos --no-enable-prefix-caching`. Baseline and DBO differ only in `--enable-dbo` and the thresholds.

### Overlap is real

Torch profiler, ~5.2 s of steady-state prefill, per rank:

| | baseline | DBO |
|---|---|---|
| comm hidden behind compute | 11.4% | **69.5%** |
| exposed (non-overlapped) comm | 1177 ms | **549 ms** |
| comm busy | 1328 ms | 1799 ms |
| compute busy | 2746 ms | 3203 ms |
| GPU busy | 3923 ms | 3753 ms |

Per-kernel, the load-bearing all-to-all ops go from `dispatch` 0.0% overlapped → 59.9% and `combine` 16.0% → 77.7%.

**But overlap does not convert 1:1.** Removing ~630 ms of exposed comm buys less than it looks like, because microbatching *adds* ~930 ms of work: twice as many all-to-all ops, where the `notify_dispatch` / `cached_notify_combine` handshakes are fixed cost per op and do not halve with the token count, plus half-size GEMMs that run less efficiently. The net of those two is what shows up end-to-end.

### End-to-end: DBO wins on large prefill

The fixed per-op cost is only amortised when each microbatch is large, which makes `--max-num-batched-tokens` the decisive knob:

| max-num-batched-tokens | microbatch size | DBO vs baseline (ISL 8192) |
|---|---|---|
| 8192 | 4096 | −8% to −13% |
| 16384 | 8192 | −1% to −2% (parity) |
| 32768 | 16384 | **+1.6%** |

At `--max-num-batched-tokens 32768`, total tok/s, multiple seeds:

| workload | baseline | DBO | gain |
|---|---|---|---|
| ISL 16384 / OSL 16, conc. 32 | 97799 | 102260 | **+4.56%** (3/3 seeds) |
| ISL 8192 / OSL 16, conc. 48 | 100423 | 101991 | **+1.56%** (4/4 seeds) |

### End-to-end: DBO loses on decode

| ISL 128 / OSL 512, conc. 256 | baseline | DBO |
|---|---|---|
| output tok/s | 4620 | 2056 (**0.44×**) |
| median TPOT | 54.6 ms | 123.7 ms |

This is structural, not a bug. The profiler shows baseline decode on this setup is **CPU-launch-bound**: GPU busy is only 11.8 ms of a 55.6 ms step (~21%), so there is no GPU bubble for DBO to fill, while microbatching doubles the launch path. The ceiling for *perfect* overlap on decode here is ~7% — far less than the launch overhead DBO adds. CUDA graphs (P3) are exactly what would change this picture, since they attack the launch bound directly.

## Open questions

**1. Does the `num_ubatches` descriptor field survive P3?** Once graphs are captured for microbatched shapes, `num_ubatches` becomes a real registry key rather than a flag that always pairs with `cg_mode=NONE`. If reviewers expect a different keying scheme, now is the time — it gets expensive to change later.

**2. Split heuristic.** The current split is the midpoint of the DP-padded token count, which all ranks agree on for free. A request-aligned split would avoid straddling requests across microbatches but needs a different agreement mechanism. Worth it?

## What's explicitly out of scope for this RFC

- LoRA, speculative decoding, pipeline parallelism, context parallelism, multimodal/encoder-decoder models, hybrid (SSM/attention) models combined with V2 DBO — all rejected at config validation, not silently degraded.
- Changing V1's DBO behavior or defaults. The only V1 file this touches is `gpu_ubatch_wrapper.py`, and only to hoist the SM-control helper into the already-shared `ubatch_utils.py` so both runners use one copy — a pure code move with no behavior change (−93/+2 lines).

### Feedback Period.

One week from posting. I'd particularly like agreement on the two open questions above, and on the two decisions that are hard to reverse later — `num_ubatches` on `BatchExecutionDescriptor`, and building per-microbatch attention metadata from pre-sliced persistent buffers rather than splitting metadata after the fact — before P3 work proceeds far enough to make them expensive to change.

### CC List.

@WoosukKwon 
@LucasWilkinson 
@njhill 
@jiangkuaixue123 

### Any Other Things.

Actual touch surface of the P0–P2 prototype (+1099 / −102 across 14 files), for reviewers sanity-checking the blast radius:

```
vllm/config/vllm.py                          (config gating, per-feature errors)
vllm/v1/worker/gpu/cudagraph_utils.py        (num_ubatches on the descriptor)
vllm/v1/worker/gpu/attn_utils.py             (per-ubatch metadata builders)
vllm/v1/worker/gpu/dp_utils.py               (DP vote for microbatching)
vllm/v1/worker/gpu/input_batch.py            (persistent per-microbatch buffers)
vllm/v1/worker/gpu/model_runner.py           (execute_model wiring, UBatchState)
vllm/v1/worker/gpu/model_states/*.py         (prepare_attn takes ubatch_idx)
vllm/v1/worker/gpu/ubatch_utils.py           (new — slicing, UBatchRunner)
vllm/v1/worker/gpu_ubatch_wrapper.py         (SM control hoisted out; pure move)
vllm/v1/worker/ubatch_utils.py               (SM control now shared)
tests/v1/worker/test_gpu_ubatch_slicing.py   (new — slicing + DP-vote units)
tests/v1/distributed/test_dbo.py             (existing DBO e2e, to be parametrized
                                              over the V1 and V2 runners)
```

Note that `ubatching.py` and the model-side `dbo_yield*` hooks are absent from that list — V2 needs no changes to the shared microbatching protocol, which was the design bet and it held.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by RyanJHamby** [comment]
Sketched a draft for the FULL-cudagraph-capture half of this RFC, built on top of #50945 (which handles the harder pre-slicing/persistent-buffer problem for eager mode) — https://github.com/RyanJHamby/vllm/commit/327520123d7a0903b28685a4282080a01dafcc2a (branch: `RyanJHamby:dbo-v2-fullgraph-capture`, based on `specture724`'s `#50945` branch).

**Status up front:** this is a design + draft diff only. It passes ast/ruff static checks; it has **not** run on a GPU, so treat the specific line-level claims below as "reasoned through," not "verified." Posting mainly to check the direction is right before I put more time into it — happy to defer entirely if `@specture724` is already planning this as a follow-up to #50945, or to fold it in there instead of a separate PR.

**Approach**, mirroring V1's `gpu_ubatch_wrapper.py::_capture_ubatches` (the one existing precedent for capturing a graph across two live microbatch threads):

- `BatchExecutionDescriptor.num_ubatches` is already threaded through `_is_compatible`/`dispatch` without being produced anywhere — `_init_candidates()` now also generates FULL-mode descriptors with `num_ubatches=2`.
- `UBatchRunner.run()` is split into `begin_capturable_run()` (spawns the microbatch threads, waits the ready barrier — pure CPU/event sync, safe to run *outside* `torch.cuda.graph()`) and a `finish()` callback (sets `cpu_wait_event`, joins, merges outputs — the part that actually needs to run *inside* the graph, since that's where the threads' compute/comm stream ops happen). `ModelCudaGraphManager`'s existing `create_forward_fn`/`forward_fn` split lines up with this almost exactly: `create_forward_fn` calls `begin_capturable_run` (outside the graph), returns a `forward_fn` that calls `finish()` (invoked once outside the graph for warmup, once inside `torch.cuda.graph()` for the real capture) — so the existing capture loop in `CudaGraphManager.capture()` didn't need to change at all.
- `_maybe_ubatch_descriptor()` (currently hardcoded to `CUDAGraphMode.NONE` with the "no CUDA graphs are captured for them yet" comment) now calls `cudagraph_manager.dispatch(..., num_ubatches=num_ubatches)`, same as the non-ubatched path — `dispatch`'s existing fallback-to-NONE already covers the "no graph captured for this shape" case for free.
- `_get_dbo_unsupported_features()` now only rejects PIECEWISE+DBO, not FULL+DBO. FULL only, on purpose — neither V1 nor #50945 support PIECEWISE+DBO, so keeping that out of scope here too.

**Open questions I don't have a confident answer to without running this:**
1. Whether `sm_control`'s enter/exit boundaries (currently: entered before thread spawn, exited inside `finish()` — i.e. inside the captured graph's Python `with` block, though after the actual `thread.join()`) are correct relative to how SM-partitioning state needs to be visible to the graph's replayed kernel launches.
2. Whether `torch.cuda.graph()`'s implicit current-stream capture (no explicit `stream=` kwarg, unlike V1's explicit `stream=compute_stream`) reliably lines up with the ubatch threads' shared `compute_stream` in this call path.

Would rather surface those now than after polishing further. If the direction looks right, next step is rebasing once #50945 lands and getting it in front of real GPU CI.

**Comment by specture724** [comment]
@RyanJHamby Appreciate your draft for full CG on DBO V2. I'm planning and drafting an PR stack for this RFC as @LucasWilkinson mentioned in the PR #50945. I'd like to rebase your implimentation on my current branch and do some tests and profiling on GPUs.
